### PR TITLE
APEX-1634 Create birthday squad

### PR DESCRIPTION
**Pull request by debrecenid**
### 1. New Feature: Birthday Squad
- Added functionality to create a birthday squad for team members celebrating their birthdays.
- Introduced a new private conversation for squad members.
- If more than 5 team members express interest, randomly select 5 members. If fewer than 3, include random team members.
- Ensured a minimum of 2 users are selected to open a conversation.

### 2. Code Updates:
- Created new files:
  - `packages/core/db/queries/getSquadMembers.ts`: Contains queries to fetch and select squad members.
  - `packages/functions/events/createBirthdaySquad.ts`: Implements the logic for creating a birthday squad, including obtaining squad members and handling errors.
  - `tests/integration/createBirthdaySquad.test.ts` and `tests/unit/createBirthdaySquad.test.ts`: Added tests to ensure the functionality works as expected.
  - `tests/utils/seedSquadJoins.ts`: Utility for seeding squad joins data for testing.
  - `packages/core/services/slack/openConversation.ts`: Updated to open a conversation for the birthday squad.

### 3. Slack Permission Update:
- Added a new Slack permission: `mpim:write` to facilitate the creation of private multi-person direct messages for the birthday squad.

## Context:
This pull request introduces a new feature to create birthday squads for team members. It leverages existing functionalities for obtaining squad members and opening conversations. Additionally, a new Slack permission, `mpim:write`, is required for creating private multi-person direct messages.

## Testing:
- Unit and integration tests have been added to ensure the correct functioning of the birthday squad feature.
- Manual testing has been conducted to validate the creation of birthday squads under various scenarios.

## Dependencies:
No new external dependencies have been introduced.

## Additional Notes:
Ensure that the new Slack permission is properly documented in the project's Slack integration documentation.

### CODE REVIEW

File: packages/functions/events/createBirthdaySquad.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  MIN_BIRTHDAY_SQUAD_SIZE,
+} from "@/utils/constants";
+import { handleEvent } from "@/utils/eventBridge/handleEvent";
+import { publishEvent } from "@/utils/eventBridge/publishEvent";
+
+export const handler = handleEvent(
+  "createBirthdaySquad",
+  async ({ team, birthdayPerson, eventId }) => {
+    try {
+      const appliedSquadMembers = await getSquadMembers(team, birthdayPerson);
```
Comment (BaDo2001) [inline packages/functions/events/createBirthdaySquad.ts:27]: Where's the logic for Kriszta to always be included?

File: packages/functions/events/createBirthdaySquad.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+      const squadMembers = [...appliedSquadMembers];
+
+      if (appliedSquadMembers.length < MIN_BIRTHDAY_SQUAD_SIZE) {
+        const randomSquadMember = await getRandomSquadMembers({
+          teamId: team,
+          usersToExclude: [birthdayPerson, ...appliedSquadMembers],
+          limit: BIRTHDAY_SQUAD_SIZE - appliedSquadMembers.length,
+        });
+        squadMembers.push(...randomSquadMember);
```
Comment (BaDo2001) [inline packages/functions/events/createBirthdaySquad.ts]:
```suggestion
        const randomSquadMembers = await getRandomSquadMembers({
          teamId: team,
          usersToExclude: [birthdayPerson, ...appliedSquadMembers],
          limit: BIRTHDAY_SQUAD_SIZE - appliedSquadMembers.length,
        });
        squadMembers.push(...randomSquadMembers);
```

File: packages/functions/events/createBirthdaySquad.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      }
+
+      if (squadMembers.length < 2) {
+        throw new Error("Error need at least 2 user to open a conversation");
+      }
+
+      const conversationId = await openConversation(squadMembers);
+
+      publishEvent("sendSquadWelcomeMessage", {
```
Comment (BaDo2001) [inline packages/functions/events/createBirthdaySquad.ts]:
```suggestion
      await publishEvent("sendSquadWelcomeMessage", {
```

File: tests/integration/createBirthdaySquad.test.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+const testSlackTeamId = import.meta.env.VITE_SLACK_TEAM_ID;
+const testUserId = import.meta.env.VITE_SLACK_USER_ID;
+const testBotUserId = import.meta.env.VITE_SLACK_BOT_USER_ID;
+
+const constants = vi.hoisted(() => ({
+  birthdayPerson: "U001",
+  eventId: "E001",
+  teamId: testSlackTeamId,
+  otherUserIds: [testUserId, testBotUserId],
+  converastionId: "CH001",
```
Comment (BaDo2001) [inline tests/integration/createBirthdaySquad.test.ts]: typo: conversationId

File: tests/integration/createBirthdaySquad.test.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+}));
+
+describe("CreateBirthdaySquad", () => {
+  beforeAll(async () => {
+    await testDb.delete(users);
+    await testDb.delete(squadJoins);
```
Comment (BaDo2001) [inline tests/integration/createBirthdaySquad.test.ts:31]: I don't think we are touching squadJoins here. It might be cleaner to create a util that wipes all tables because it's a bit of a mess right now. What do you think?

File: tests/unit/createBirthdaySquad.test.ts [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+import { sendMockSqsMessage } from "@/testUtils/unit/sendMockSqsMessage";
+
+dayjs.extend(utc);
+
+const constants = vi.hoisted(() => ({
+  birthdayPerson: "U001",
+  eventId: "E001",
+  teamId: "T001",
+  otherUserIds: ["U002", "U003", "U004", "U005", "U006"],
+  converastionId: "CH001",
```
Comment (BaDo2001) [inline tests/unit/createBirthdaySquad.test.ts]: same typo as above

File: tests/unit/createBirthdaySquad.test.ts [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+}));
+
+vi.mock("@/services/slack/openConversation", async () => ({
+  openConversation: vi.fn().mockResolvedValue(constants.converastionId),
+}));
+
+describe("CreateBirthdaySquad", () => {
+  let openConversationMock: Mock;
```
Comment (BaDo2001) [inline tests/unit/createBirthdaySquad.test.ts]: Do we need this mock? As far as I see we are always returning the same conversationId so the mock above is enough.

File: tests/unit/createBirthdaySquad.test.ts [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+
+    openConversationMock.mockResolvedValueOnce(constants.converastionId);
+
+    await sendMockSqsMessage("createBirthdaySquad", event, handler);
+
+    expect(getSquadMembersSpy).toHaveBeenCalledWith(
+      constants.teamId,
+      constants.birthdayPerson,
+    );
+  });
```
Comment (BaDo2001) [inline tests/unit/createBirthdaySquad.test.ts]: This test doesn't seem too meaningful for me. We should be testing cases when this function returns something but just the fact that it gets called holds little value.

File: tests/unit/createBirthdaySquad.test.ts [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+  afterEach(async () => {
+    vi.clearAllMocks();
+  });
+
+  it("Should not call openConverstaion", async () => {
+    const event = {
+      birthdayPerson: constants.birthdayPerson,
+      eventId: constants.eventId,
+      team: constants.teamId,
+    } satisfies Events["createBirthdaySquad"];
```
Comment (BaDo2001) [inline tests/unit/createBirthdaySquad.test.ts]: I can see this before every test so we should move it to a constant as is.

File: tests/unit/createBirthdaySquad.test.ts [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+    await sendMockSqsMessage("createBirthdaySquad", event, handler);
+
+    expect(getRandomSquadMembersSpy).toHaveBeenCalledWith({
+      teamId: constants.teamId,
+      usersToExclude: [constants.birthdayPerson],
+      limit: BIRTHDAY_SQUAD_SIZE,
+    });
+
+    expect(PutEventsCommand).not.toBeCalled();
+    expect(eventBridge.send).not.toBeCalled();
```
Comment (BaDo2001) [inline tests/unit/createBirthdaySquad.test.ts:103]: This test is the same as above just expecting different things not to be called. I think we should merge the two.

File: tests/unit/createBirthdaySquad.test.ts [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+    await sendMockSqsMessage("createBirthdaySquad", event, handler);
+
+    expect(getSquadMembersSpy).toHaveBeenCalledWith(
+      constants.teamId,
+      constants.birthdayPerson,
+    );
+  });
+});
+
+describe("Final squad size is less then 2", () => {
```
Comment (BaDo2001) [inline tests/unit/createBirthdaySquad.test.ts:64]: I am not 100% sure on this, but imo different test suits should be for different things. Here its the same function so I think all tests should be in one describe block.

File: tests/unit/createBirthdaySquad.test.ts [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+      birthday: dayjs.utc().toDate(),
+    });
+
+    await testDb.insert(users).values(
+      constants.otherUserIds.map((userId) => ({
+        id: userId,
+        teamId: constants.teamId,
+        birthday: dayjs.utc().toDate(),
+      })),
+    );
```
Comment (BaDo2001) [inline tests/unit/createBirthdaySquad.test.ts:143]: I would move this inside the seeding too.

File: tests/unit/createBirthdaySquad.test.ts [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+
+    insertedSquadMembers = await seedSquadJoins(
+      constants.birthdayPerson,
+      constants.teamId,
+      constants.otherUserIds,
+      3,
+    );
+  });
+
+  afterEach(async () => {
```
Comment (BaDo2001) [inline tests/unit/createBirthdaySquad.test.ts]: Shouldn't we also have a beforeAll clean? If another test doesn't clean up properly or something then we might not start with an empty db.

File: tests/unit/createBirthdaySquad.test.ts [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+    );
+
+    openConversationMock.mockResolvedValueOnce(constants.converastionId);
+
+    await sendMockSqsMessage("createBirthdaySquad", event, handler);
+
+    expect(getRandomSquadMembersSpy).not.toHaveBeenCalled();
+  });
+
+  it("Should call openConversation", async () => {
```
Comment (BaDo2001) [inline tests/unit/createBirthdaySquad.test.ts:170]: Once again I think these 3 test cases could be merged into one as they are testing the same flow.

File: tests/unit/createBirthdaySquad.test.ts [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+    expect(getRandomSquadMembersSpy).toHaveBeenCalledWith({
+      teamId: constants.teamId,
+      usersToExclude: [constants.birthdayPerson],
+      limit: BIRTHDAY_SQUAD_SIZE,
+    });
+
+    expect(openConversation).not.toBeCalled();
+  });
+
+  it("Should not call publishEvent", async () => {
```
Comment (BaDo2001) [inline tests/unit/createBirthdaySquad.test.ts]: To me this is not informative enough because I don't know what kind of event we were expecting.

File: tests/unit/createBirthdaySquad.test.ts [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+  beforeEach(() => {
+    openConversationMock = openConversation as Mock;
+    eventBridge = new EventBridgeClient();
+  });
+
+  afterEach(async () => {
+    vi.clearAllMocks();
+  });
+
+  it("Should not call openConverstaion", async () => {
```
Comment (BaDo2001) [inline tests/unit/createBirthdaySquad.test.ts]:
```suggestion
  it("Should not call openConversation", async () => {
```

File: tests/unit/createBirthdaySquad.test.ts [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+      usersToExclude: [constants.birthdayPerson],
+      limit: BIRTHDAY_SQUAD_SIZE,
+    });
+
+    expect(PutEventsCommand).not.toBeCalled();
+    expect(eventBridge.send).not.toBeCalled();
+  });
+});
+
+describe("Applied squad members is less the 3", () => {
```
Comment (BaDo2001) [inline tests/unit/createBirthdaySquad.test.ts]:
```suggestion
describe("3 or less squad members applied", () => {
```

File: tests/unit/createBirthdaySquad.test.ts [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+
+    expect(getRandomSquadMembersSpy).toHaveBeenCalledWith({
+      teamId: constants.teamId,
+      usersToExclude: [constants.birthdayPerson],
+      limit: BIRTHDAY_SQUAD_SIZE,
+    });
+  });
+});
+
+describe("Applied squad members is 3 or more", () => {
```
Comment (BaDo2001) [inline tests/unit/createBirthdaySquad.test.ts]:
```suggestion
describe("3 or more squad members applied", () => {
```

Comment (BaDo2001) [reply tests/unit/createBirthdaySquad.test.ts]: The one above is 3 or less and this one is 3 or more. Seems odd because test cases shouldn't be intersecting.

**Review by BaDo2001 — CHANGES_REQUESTED** [review]
Good overall, few things:
- The admin part is not covered (i.e either Kriszta or Máté is included)
- The unit tests could be organized better
  - One test suite as its all for one function
  - If two tests are only different in their outcomes, they could be merged (has to be checked if they can fail separately, i.e if test1 and test2 will always pass or fail together than they can probably be merged together.)
  - Some typos and unnecessary mocks
  - We should assert the people who got put into the conversation are the ones we were expecting
- No integration test for squad join

File: tests/integration/createBirthdaySquad.test.ts (hunk 5 above)
Comment (debrecenid) [reply tests/integration/createBirthdaySquad.test.ts:31]: I think thats fine if we just take care what we want to clean up. I removed the squadJoins cleanup from here

File: tests/unit/createBirthdaySquad.test.ts (hunk 12 above)
Comment (debrecenid) [reply tests/unit/createBirthdaySquad.test.ts:143]: After I merged them a bit, which was not what you suggested, but all was separate before,with multiple seeding, then I merged them together under some describe, then it been used only one place, along with the seeding, so actually the seeding is also not necessary.

**Comment by BaDo2001** [comment]
@debrecenid Also, could you please merge main into your branch and add a manual testing endpoint?

File: packages/functions/events/createBirthdaySquad.ts [hunk 19]
```diff
@@ -13,27 +15,36 @@ import { publishEvent } from "@/utils/eventBridge/publishEvent";
...
     try {
       const appliedSquadMembers = await getSquadMembers(team, birthdayPerson);
 
       const squadMembers = [...appliedSquadMembers];
 
       if (appliedSquadMembers.length < MIN_BIRTHDAY_SQUAD_SIZE) {
-        const randomSquadMember = await getRandomSquadMembers({
+        const randomSquadMembers = await getRandomSquadMembers({
           teamId: team,
           usersToExclude: [birthdayPerson, ...appliedSquadMembers],
```
Comment (BaDo2001) [inline packages/functions/events/createBirthdaySquad.ts]: Shouldn't we exclude Kriszta / Máté so we are always pulling other members in?

File: tests/integration/createBirthdaySquad.test.ts [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+
+      await testDb.insert(users).values(
+        constants.otherUserIds.map((userId) => ({
+          id: userId,
+          teamId: constants.teamId,
+          birthday: dayjs.utc().toDate(),
+        })),
+      );
+
+      const eventId = "CS1_" + Date.now().toString();
```
Comment (BaDo2001) [inline tests/integration/createBirthdaySquad.test.ts]:
```suggestion
      const eventId = "CS2_" + Date.now().toString();
```

**Review by BaDo2001 — CHANGES_REQUESTED** [review]

File: packages/functions/events/createBirthdaySquad.ts (hunk 19 above)
Comment (debrecenid) [reply packages/functions/events/createBirthdaySquad.ts]: 🤯

File: tests/unit/createBirthdaySquad.test.ts (hunk 18 above)
Comment (BaDo2001) [reply tests/unit/createBirthdaySquad.test.ts]: @debrecenid This is still not clear to me.

**Review by BaDo2001 — CHANGES_REQUESTED** [review]

**Review by BaDo2001 — APPROVED** [review]
